### PR TITLE
Remove patch version check for plugin compatibility

### PR DESCRIPTION
**Pull request by abseth-amzn**
Signed-off-by: Abhilasha Seth <abseth@amazon.com>

### Description
As a first step towards supporting semVer range of compatible versions for plugins, we are starting with relaxing the patch version check (for default compatibility range). 

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/1707

### Check List
- [ ] New functionality includes testing.
  - [X] New tests pass
- [ ] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dblock — APPROVED** [review]

File: server/src/test/java/org/opensearch/plugins/PluginsServiceTests.java [hunk 1]
```diff
@@ -717,6 +718,19 @@ public void testIncompatibleOpenSearchVersion() throws Exception {
         assertThat(e.getMessage(), containsString("was built for OpenSearch version 6.0.0"));
     }
 
+    public void testPluginDifferingWithCoreForPatchVersion() throws Exception {
+        // Get a version for plugin to use that is just before the core's version.
+        // We will use the plugin version for test only if it differs in patch version
+        Version pluginVersion = VersionUtils.getPreviousVersion(Version.CURRENT);
+        if (pluginVersion.major != Version.CURRENT.major || pluginVersion.minor != Version.CURRENT.minor) {
```
Comment (dblock) [inline server/src/test/java/org/opensearch/plugins/PluginsServiceTests.java:725]: This doesn't look like a predictable test. Write tests that have predictable versions of the hypothetical plugin and cover all scenarios.

**Comment by dblock** [comment]
We need to discuss whether relaxing plugin compatibility versions is a breaking change. @saratvemulapalli wdyt?

**Review by saratvemulapalli — COMMENTED** [review]
Lets also take care of adding tests to: https://github.com/opensearch-project/OpenSearch/blob/main/distribution/tools/plugin-cli/src/test/java/org/opensearch/plugins/InstallPluginCommandTests.java

This internally uses `PluginsService.java` `verifyCompatibility`.

**Comment by saratvemulapalli** [comment]
> We need to discuss whether relaxing plugin compatibility versions is a breaking change. If it is, then we need this behavior to be configurable and default to `false`.
> 
> @saratvemulapalli wdyt?

Good question, looking at Semver[1] `MINOR version when you add functionality in a backwards compatible manner`. 
This feature is backward compatible since we are relaxing patch version, our installation APIs will not break.

That said, for developers who are using this feature we really want to "caution" that its only for users "who know what they are doing". Because there are risks associated with relaxing patch versions and might up into runtime problems. 
Gating it by a feature flag would help lower the risk. 

@abseth-amzn we could add another parameter `plugin-descriptor.properties`[1] file and enable devs to configure this feature and set it false by default if it doesn't exist and open an issue in https://github.com/opensearch-project/documentation-website/ to document details of this feature. 

[1] https://github.com/opensearch-project/OpenSearch/blob/main/buildSrc/src/main/resources/plugin-descriptor.properties

**Comment by reta** [comment]
@dblock @saratvemulapalli I think we could not relax the versioning check in one directional way, the plugin should have a way to set the OpenSearch core baseline it need to works with. There are at least 2 examples why its needed:
 - bug that is present in one version and prevents plugin from functioning properly (example, 2.6.1 - OK,  2.6.2 - BUG, 2.6.3 - BUG FIXED), they way plugin developers could circumvent it right now - just skip 2.6.2 plugin release altogether
 - new plugin interface / method introduced (in a fully backward compatible fashion, non-breaking change in patch version, examples are CodecServiceProvider) - the plugin just won't with any previous version of core

**Comment by saratvemulapalli** [comment]
> bug that is present in one version and prevents plugin from functioning properly (example, 2.6.1 - OK, 2.6.2 - BUG, 2.6.3 - BUG FIXED), the way plugin developers could circumvent it right now - just skip 2.6.2 plugin release altogether

Thanks @reta, I hope this should answer your question: https://github.com/opensearch-project/OpenSearch/issues/1707#issuecomment-1459207057
`Plugin able to define range of patch version compatibility i.e >2.5.1, <2.5.7`, for some patch version upgrades on OpenSearch few plugins are not impacted and this feature enables them to completely ignore the check and not ship a new version. 

> new plugin interface / method introduced (in a fully backward compatible fashion, non-breaking change in patch version, examples are CodecServiceProvider) - the plugin just won't work with any previous version of core

Absolutely, if the plugin chooses to use the new interface it will bump its own patch version and will define it minimum compatibility version of OpenSearch when the new interface is added.

Would that solve the problem? Or am I missing something here.

Also this made me think, assuming the plugin shipped with `2.7.0` of OpenSearch defining it compatibility `>2.7`. 
When a new patch version of OpenSearch releases `2.7.2`, which doesnt work with plugin 2.7.0.0. How does the plugin tell customers that it doesn't work with `2.7.2` since its already shipped and doesn't know about the future releases.

**Comment by reta** [comment]
> Would that solve the problem? Or am I missing something here.

Thanks @saratvemulapalli , the discussion are addressing the concerns, but I don't see that reflected in the pull request 

> How does the plugin tell customers that it doesn't work with 2.7.2 since its already shipped and doesn't know about the future releases.

This is somewhat tricky, but the suggestion from the comment you referred to (https://github.com/opensearch-project/OpenSearch/issues/1707#issuecomment-1459207057) gives the hint: the plugin developers could not modify the released plugins, but they do have a choice to keep lockstep releases (basically, just limiting to exact release, `>= 2.6.1`, `< 2.6.2`)

**Comment by dblock** [comment]
I think we're all in agreement of what we would like in this PR.

1. New feature: plugins have a way to explicitly define a semver version range.
2. Server will match the semver version range defined by the plugin when available.
3. No changes to existing plugins that don't want this feature.

**Comment by dblock** [comment]
DCO needs fixing please, and make sure gradle checks pass, will review when 🟢 again

File: server/src/test/java/org/opensearch/plugins/PluginsServiceTests.java [hunk 2]
```diff
@@ -717,6 +718,19 @@ public void testIncompatibleOpenSearchVersion() throws Exception {
...
+            // Cannot fail the test as there is no version that differs with core's patch version
+            // but has the same major and minor version as core. Not a valid test in that case.
+            return;
+        }
+        PluginInfo info = new PluginInfo("my_plugin", "desc", "1.0", pluginVersion, "1.8", "FakePlugin", Collections.emptyList(), false);
+        PluginsService.verifyCompatibility(info);
```
[the remark this answers is not in the document]
Comment (abseth-amzn) [reply server/src/test/java/org/opensearch/plugins/PluginsServiceTests.java:806]: Added tests that call isPluginVersionCompatible directly and assert.

File: server/src/test/java/org/opensearch/plugins/PluginsServiceTests.java (hunk 1 above)
Comment (abseth-amzn) [reply server/src/test/java/org/opensearch/plugins/PluginsServiceTests.java]: Added predictable tests.

**Comment by saratvemulapalli** [comment]
Re-opening this PR. This is a needed change.

**Review by dblock — CHANGES_REQUESTED** [review]
I'm opposed to this change as is (a boolean that says that a plugin is compatible with patch versions of OpenSearch). I tried to say it in https://github.com/opensearch-project/OpenSearch/pull/6837#pullrequestreview-1437147520 above. The tl;dr is:

1. If a plugin is authored for OpenSearch 2.8+, a boolean will say it's also compatible with 2.7, but it will not work.
2. A boolean cannot evolve into a semver range in the future, so it's not a good foundation to build upon.

I would be ok with a semver compatible range that can be declared by a plugin, even if only a subset of it is implemented. As a plugin author I'd like to write `">= 2.5"`, `"~> 2.5"`, etc., in the plugin manifest. I would be ok if the implementation doesn't support more complex syntax in v1, but I want to make sure we can add support for any semver syntax in the future without having to change the interface.

**Comment by dblock** [comment]
> I'm opposed to this change as is (a boolean that says that a plugin is compatible with patch versions of OpenSearch).

@abseth-amzn Please address my comment.

**Comment by abseth-amzn** [comment]
Closing in favor of https://github.com/opensearch-project/OpenSearch/pull/11441

**Pull request closed without merging**